Wrap() with struct (not pointer) fails silently

Passing a struct to `Wrap` should work, according to the docs, however it fails silently. It accepts the struct and returns a value but getting any attributes from the wrapped value returns its zero-value (or something along those lines).

To see this, in the `wrapper_test.go` test `TestWrapper` change this line: 

	wrap1 := Wrap(res1)

to 

	wrap1 := Wrap(*res1)

The test fails, though it should still pass. There are many errors:

```
--- FAIL: TestWrapper (0.00s)
    /Documents/Code/jsonapi/wrapper_test.go:67: 
        	Error Trace:	wrapper_test.go:67
        	Error:      	Not equal: 
        	            	expected: "res123"
        	            	actual  : ""
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-res123
        	            	+
        	Test:       	TestWrapper
        	Messages:   	id
    /Documents/Code/jsonapi/wrapper_test.go:71: 
        	Error Trace:	wrapper_test.go:71
        	Error:      	Not equal: 
        	            	expected: "res123"
        	            	actual  : "another-id"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-res123
        	            	+another-id
        	Test:       	TestWrapper
        	Messages:   	set id

//.... many more errors...
```

I still think there is something wrong with wrapped structs (non-pointer). They return `""` for most gets.

Take the test you added and change it to be a non-pointer, and just assert that what you put in is what you get out:

```
func TestWrapStruct(t *testing.T) {
	assert := assert.New(t)

	res1 := mockType1{
		ID:  "res123",
		Str: "a_string",
	}

	wrap1 := Wrap(res1)

	// ID, type, field
	id, typ := wrap1.IDAndType()
	assert.Equal(res1.ID, id, "id")
	assert.Equal("mocktypes1", typ, "type")
	assert.Equal(res1.Str, wrap1.Get("str"), "str field")
}
```

This should pass, if I understand the idea of wrappers correctly, but I get this:

```
--- FAIL: TestWrapStruct (0.00s)
    /Code/jsonapi/wrapper_test.go:53: 
        	Error Trace:	wrapper_test.go:53
        	Error:      	Not equal: 
        	            	expected: "res123"
        	            	actual  : ""
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-res123
        	            	+
        	Test:       	TestWrapStruct
        	Messages:   	id
    /Code/jsonapi/wrapper_test.go:55: 
        	Error Trace:	wrapper_test.go:55
        	Error:      	Not equal: 
        	            	expected: "a_string"
        	            	actual  : ""
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-a_string
        	            	+
        	Test:       	TestWrapStruct
        	Messages:   	str field
FAIL
FAIL	github.com/mfcochauxlaberge/jsonapi	0.013s
FAIL
```

--- a/wrapper.go
+++ b/wrapper.go
@@ -26,22 +26,34 @@ type Wrapper struct {
 // Wrap wraps v (a struct or a pointer to a struct) and returns a Wrapper that
 // can be used as a Resource to handle the given value.
 //
-// If v is not a pointer, the changes applied to the Wrapper object won't affect
-// the underlying object (which will be a new instance of v's type).
+// Changes made to the Wrapper object (through Set for example) will be applied
+// to v.
+//
+// If v is not a pointer, a copy is made and v won't be modified by the wrapper.
 func Wrap(v interface{}) *Wrapper {
 	val := reflect.ValueOf(v)
 
-	if val.Kind() != reflect.Ptr {
+	switch {
+	case val.Kind() != reflect.Ptr:
 		if val.Kind() != reflect.Struct {
 			panic(errors.New("jsonapi: value has to be a pointer to a struct"))
 		}
 
-		val = reflect.New(val.Type())
-	} else if val.Elem().Kind() != reflect.Struct {
-		panic(errors.New("jsonapi: value has to be a pointer to a struct"))
-	}
+		newVal := reflect.New(val.Type()).Elem()
 
-	val = val.Elem()
+		for i := 0; i < newVal.NumField(); i++ {
+			f := newVal.Field(i)
+			if f.CanSet() {
+				f.Set(val.Field(i))
+			}
+		}
+
+		val = newVal
+	case val.Elem().Kind() != reflect.Struct:
+		panic(errors.New("jsonapi: value has to be a pointer to a struct"))
+	default:
+		val = val.Elem()
+	}
 
 	err := Check(val.Interface())
 	if err != nil {

--- a/wrapper_test.go
+++ b/wrapper_test.go
@@ -41,7 +41,7 @@ func TestWrap(t *testing.T) {
 func TestWrapStruct(t *testing.T) {
 	assert := assert.New(t)
 
-	res1 := &mockType1{
+	res1 := mockType1{
 		ID:  "res123",
 		Str: "a_string",
 	}
@@ -54,13 +54,20 @@ func TestWrapStruct(t *testing.T) {
 	assert.Equal("mocktypes1", typ, "type")
 	assert.Equal(res1.Str, wrap1.Get("str"), "str field")
 
-	// Modyfing the wrapper does not modify
+	// Modifying the wrapper does not modify
 	// the original value.
-	wrap1.SetID("another-id")
+	wrap1.SetID("another_id")
 	id, _ = wrap1.IDAndType()
-	wrap1.Set("str", "another_string")
 	assert.Equal("another_id", id, "type")
-	assert.Equal(res1.Str, wrap1.Get("str"), "str field")
+
+	wrap1.Set("str", "another_string")
+	assert.Equal("another_string", wrap1.Get("str"), "str field")
+
+	// Modifying the original value does
+	// not modify the wrapper.
+	res1.Str = "new_string"
+	assert.NotEqual(res1.Str, wrap1.Get("str"), "str field")
+	assert.Equal("another_string", wrap1.Get("str"), "str field")
 }
 
 func TestWrapper(t *testing.T) {

--- a/wrapper_test.go
+++ b/wrapper_test.go
@@ -38,6 +38,30 @@ func TestWrap(t *testing.T) {
 	}, "panic when not a valid struct")
 }
 
+func TestWrapStruct(t *testing.T) {
+	assert := assert.New(t)
+
+	res1 := &mockType1{
+		ID:  "res123",
+		Str: "a_string",
+	}
+
+	wrap1 := Wrap(res1)
+
+	// ID, type, field
+	id, typ := wrap1.IDAndType()
+	assert.Equal(res1.ID, id, "id")
+	assert.Equal("mocktypes1", typ, "type")
+	assert.Equal(res1.Str, wrap1.Get("str"), "str field")
+
+	// Modyfing the wrapper does not modify
+	// the original value.
+	wrap1.SetID("another-id")
+	wrap1.Set("str", "another_string")
+	assert.Equal("mocktypes1", typ, "type")
+	assert.Equal(res1.Str, wrap1.Get("str"), "str field")
+}
+
 func TestWrapper(t *testing.T) {
 	assert := assert.New(t)
 

--- a/wrapper_test.go
+++ b/wrapper_test.go
@@ -57,8 +57,9 @@ func TestWrapStruct(t *testing.T) {
 	// Modyfing the wrapper does not modify
 	// the original value.
 	wrap1.SetID("another-id")
+	id, _ = wrap1.IDAndType()
 	wrap1.Set("str", "another_string")
-	assert.Equal("mocktypes1", typ, "type")
+	assert.Equal("another_id", id, "type")
 	assert.Equal(res1.Str, wrap1.Get("str"), "str field")
 }
 

